middle_end/flambda2/tests/algorithms/patricia_tree_tests.ml

```diff
@@ -0,0 +1,915 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*           Mark Shinwell and Leo White, Jane Street Europe              *)
+(*                                                                        *)
+(*   Copyright 2015--2022 Jane Street Group LLC                           *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+[@@@ocaml.warning "+a-4-30-40-41-42"]
+
+module Key = struct
+  type t = int
+
+  let print ppf i =
+    let chunk n = (i lsr (n * 16)) land 0xffff in
+    Format.fprintf ppf "%04x_%04x_%04x_%04x" (chunk 3) (chunk 2) (chunk 1)
+      (chunk 0)
+end
+
+module Value = struct
+  type t =
+    | A
+    | B
+
+  let equal t1 t2 = match t1, t2 with A, A | B, B -> true | _ -> false
+
+  let compare t1 t2 =
+    match t1, t2 with A, A | B, B -> 0 | A, B -> -1 | B, A -> 1
+
+  let print ppf t =
+    Format.pp_print_string ppf (match t with A -> "A" | B -> "B")
+end
+
+module Tree = Flambda2_algorithms.Patricia_tree.Make (Key)
+module Set = Tree.Set
+module Map = Tree.Map
+
+let ( <=> ) = Bool.equal
+
+let ( = ) = Int.equal
+
+let raises_not_found f x =
+  match f x with _ -> false | exception Not_found -> true
+
+let option_of_not_found f x =
+  match f x with y -> Some y | exception Not_found -> None
+
+module type Value = sig
+  type t
+
+  val equal : t -> t -> bool
+
+  val compare : t -> t -> int
+end
+
+let reflexive compare a = compare a a = 0
+
+let antisymmetric compare a b = compare a b = ~-(compare b a)
+
+let transitive compare a b c =
+  let cmp_a_b = compare a b in
+  compare b c != cmp_a_b || compare a c = cmp_a_b
```

Comment:
Hmm ... the lack of short-circuiting bothers me (though I admit that neither this nor the other "not or" instance actually _requires_ short-circuit evaluation). I made a version of `(==>)` (from Core syntax) that takes a thunk.

Reply:
I like the change a lot for clarity, but it does make me worry about test coverage... that said, I'm sort of at a "the perfect is the enemy of the good" place wrt that for this stuff, I think you're getting a lot of confidence and that's worth a lot.

Reply:
Well, even in this case, if we neglect the equal case altogether (and it's hard to imagine breaking that case here and not elsewhere), ~50% of cases should pass the implication, since 1 and -1 are equally likely as the values of `compare a b` and `compare b c`. So coverage should be pretty good.